src/nvim/api/buffer.c

```diff
@@ -636,6 +637,185 @@ ArrayOf(Integer, 2) buffer_get_mark(Buffer buffer, String name, Error *err)
   return rv;
 }
 
+/// Return a tuple (row,col) representing the position of the named mark
+///
+/// @param buffer The buffer handle
+/// @param name The mark's name
+/// @param[out] err Details of an error that may have occurred
+/// @return The (row, col) tuple
+ArrayOf(Integer, 2) mark_index(Buffer buffer, String name, Error *err)
+{
+  Array rv = ARRAY_DICT_INIT;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  if (!buf) {
+    return rv;
+  }
+
+  if (name.size < 1) {
+    api_set_error(err, Validation, _("Mark name must be at least one char"));
+    return rv;
+  }
+
+  pos_T *pos = extmark_index(buf, name.data);
+  if (pos == NULL) {
+    api_set_error(err, Validation, _("Invalid mark name"));
+    return rv;
+  }
+
+  ADD(rv, INTEGER_OBJ(pos->lnum));
+  ADD(rv, INTEGER_OBJ(pos->col));
+
+  return rv;
+}
+
+/// Returns an ordered tuple of mark names in the buffer
+///
+/// @param buffer The buffer handle
+/// @param[out] err Details of an error that may have occurred
+/// @return Tuple of mark names
+ArrayOf(String) mark_names(Buffer buffer, Error *err)
+{
+  Array rv = ARRAY_DICT_INIT;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  if (!buf) {
+    return rv;
+  }
+
+  ExtmarkNames *all_mark_names = extmark_names(buf);
+  // TODO yield results i.e iterator?
+  cstr_t mark_name;
+  for (size_t i = 0; i < kv_size(*all_mark_names); i++) {
+    mark_name = kv_A(*all_mark_names, i);
+    ADD(rv, STRING_OBJ(cstr_to_string(mark_name)));
+  }
+
+  return rv;
+}
+
+//TODO allos position index or string index
+//    If the index is in numeric form, the method returns the first mark at that position. If the index is a mark, the method returns the next mark following that mark, which may be at the same numerical position.
+
+/// Returns the position of the mark following the given mark
+/// name; if there are no following marks, the method returns
+/// an empty list.
+///
+/// @param buffer The buffer handle
+/// @param name The mark's name
+/// @param[out] err Details of an error that may have occurred
+/// @return The (row, col) tuple
+ArrayOf(Integer, 2) mark_next(Buffer buffer, String name, Error *err)
```

Comment:
While this is a good api for consumers (vimscript, lua) within the nvim process, for remote plugins it could be too slow as it requires a roundtrip (two context switches) per mark it passes by. I would imagine a function `nvim_buf_get_extmarks(buffer, start, end, which)` which would return marks in some range `[start,end]` where the boundaries ideally could be either explicit positions (including the start/end of the entire buffer), or the name of marks, and one could request to either get the N first or N last or all marks in that range. (which will cover the ability of these functions)